text/0000-ternary.md

```diff
@@ -0,0 +1,238 @@
++ Feature name: (fill me in with a unique ident, `my_awesome_feature`)
++ Start date: (fill me in with today's date, YYYY-MM-DD)
++ RFC PR: [iotaledger/bee-rfcs#0000](https://github.com/iotaledger/bee-rfcs/pull/0000)
++ Bee issue: [iotaledger/bee#0000](https://github.com/iotaledger/bee/issues/0000)
+
+# Summary
+
+In this RFC we introduce both the API and implementation of the `bee-ternary`
+crate, a general-purpose ternary manipulation crate written in Rust that may be
+used to handle ternary data in the IOTA ecosystem.
+
+# Motivation
+
+Ternary is a fundamental part of the IOTA ecosystem and is used throughout, even
+for fundamental features like wallet addresses and transaction identification.
+
+Manipulating ternary in an ergonomic manner on top of existing binary-driven
+code can require complexity and, up until now, comes either with a lot of
+boilerplate code, implementations that are difficult to verify as correct, or a
+lack of features.
+
+`bee-ternary` seeks to become the canonical ternary crate in the Rust IOTA
+ecosystem by providing an API that is efficient, ergonomic, and featureful all
+at once.
+
+We hope that `bee-ternary` will allow IOTA developers to more easily write code
+that works with ternary and allow the simplification of existing codebases
+through use of the API.
+
+Broadly, there are 3 main benefits things that `bee-ternary` aims to provide
+over use-specific implementations of ternary manipulation code.
+
+- **Ergonomics**: The API should be trivial to use correctly and should
+naturally guide developers towards correct and efficient solutions.
+
+- **Features**: The API should provide a powerful set of core features that can
+be used together to cover most developer requirements. Examples of such features
+include multiple encoding schemes, binary/ternary conversion, (de)serialization,
+and tryte string de/encoding.
+
+- **Performance**: The API should allow 'acceptably' efficient manipulation of
+ternary data. This is difficult to ruggedly define, and implementation details
+may change later, but broadly we intend to introduce an API that doesn't
+inherently inhibit high performance through poor design choices.
+
+# Detailed design
+
+`bee-ternary` is designed to be extensible and is built on top of a handful of
+powerful abstraction types. Below we list the features of the API and summarise
+a few of the core API objects.
+
+## Features
+
+- Efficient manipulation of ternary buffers (trits and trytes)
+- Multiple encoding schemes
+- Extensible design that allows it to sit on top of existing data structures,
+avoiding unnecessary allocation and copying
+- A plethora of utility functions to allow for easy manipulation of ternary data
+- Zero-cost conversion between trit and tryte formats
+
+## Key Types & Traits
+
+### `Trits`
+
+```rust
+struct Trits<T: RawEncoding> { ... }
```

Comment:
Maybe we should move the `RawEncoding` trait before this `struct`? Then it might be easier to understand why `Trits` generalizes over it. What do you think?

Reply:
I'm against this largely because it's not a common thing for an API user to interact with. As of yet, no code in the rest of Bee explicitly uses this API feature and so I'm reluctant to put it first since it may confuse users.

Reply:
okay. I understand. Maybe an introductory sentence then. Because atm when the reader reaches this point of the RFC (and understood everything so far), he is now confronted with a `struct` definition that uses a still undefined trait bound `RawEncoding`.

Reply:
Will address this with an explanation